apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/observer/ShardObserver.java

```diff
@@ -45,33 +47,64 @@ public void onTrimDone(@Observes @Sync TrimData trimData) {
 //do it async anyway because we have to exit from trim and unlock it       
         tryCreateShardAsync(trimData.getTrimHeight(), trimData.getBlockchainHeight());
     }
-
-    public CompletableFuture<MigrateState> tryCreateShardAsync(int lastTrimBlockHeight, int blockchainHeight) {
-        CompletableFuture<MigrateState> completableFuture = null;
-        HeightConfig currentConfig = blockchainConfig.getCurrentConfig();
+    
+    private boolean isShardingEnabled(HeightConfig currentConfig){
         boolean isShardingOff = propertiesHolder.getBooleanProperty("apl.noshardcreate", false);
         boolean shardingEnabled = currentConfig.isShardingEnabled();
         log.debug("Is sharding enabled ? : '{}' && '{}'", shardingEnabled, !isShardingOff);
+        return shardingEnabled && !isShardingOff;
+    }
+    
+    //TODO: check and debug
+    private boolean isShardAlreadyCreatedOrInProgress(int  lastTrimBlockHeight, HeightConfig currentConfig){
+        Shard shard = shardService.getLastShard();
+        if(shard==null){
+            return false;
+        }        
+        int lastShardHeight = shard.getShardHeight();
+        int nextShardHeight = lastShardHeight+currentConfig.getShardingFrequency();
```

Comment:
You should use shardFrequency calculated in 'isTimeForShard' method.

Reply:
done